### ISSUE TITLE
How to use ace/tern with ace-builds?

### ISSUE THREAD

**Issue by jparismorgan** [ajaxorg/ace-builds#218]
Hi there. I am trying to use `tern` with ace-builds, but it is not working, as I'm getting errors when doing:
```
ace.config.loadModule('ace/ext/language_tools', () => {
      ace.config.loadModule('ace/ext/tern', () => {
            ...
```
<img width="797" alt="Screen Shot 2021-09-15 at 10 02 00 PM" src="https://user-images.githubusercontent.com/1396242/133552478-cc1a5989-170c-4e97-bf76-0a9dae44c364.png">

Things I have seen:
* This repo does not have `ext-tern.js` in it: https://github.com/ajaxorg/ace-builds/tree/master/src-noconflict
* The `ace-tern` repo does have `ext-tern.js` in it: https://github.com/sevin7676/Ace.Tern/tree/master/ace-builds/src-noconflict
* At the same time I found this project online that seems to use `ace-builds` and `tern` together successfully: https://github.com/tmone/incentive/blob/64a89587b002e9103f41a581165bb2ee214d0e79/NC.VS/Resource/lib/Ace_Tern/demo.html#L51

So, curious, should I be able to use `tern` with `ace-builds`?

cc @sevin7676

**Comment by JSmithOner** [comment]
@jparismorgan  Did you manage to solve this issue? I'm using angular 11 and managed to make my app load the module  but appart from loading and executing the worker nothing happens. Best.

**Comment by jparismorgan** [comment]
@JSmithOner I did not. Could you share your progress in getting it loaded? Could perhaps try from there and see if I can make any progress.

**Comment by JSmithOner** [comment]
Thanks for your quick reply. Sure.
I've used file in this update repo [JSCFEditor](https://github.com/g--o/JSCFEditor/tree/0a2cb0c6d48ad2c00ea50101aae107e8e9ebdbfc/src/src-noconflict)
just place `ext-tern.js` and `worker-tern.js` into the folder `node_modules/ace-builds/src-noconflict` then in your `angular.json` create sort of a symlink as so:

`    ...        
    "assets": [
              "src/favicon.ico",
              "src/assets",
              {
                "glob": "**/*",
                "input": "./node_modules/ace-builds/src-noconflict/",
                "output": "/assets/js/ace",
                "followSymlinks": true
              }
            ],
    ...
`

then in your service or component add:

`  
import * as ace from "ace-builds";
...
initEditor ( element:ElementRef<HTMLElement>) {

    this.aceEditor = ace.edit( element.nativeElement );

    ace.config.set('basePath','assets/js/ace/');
    ace.config.set("fontSize", "14px");

    this.aceEditor.session.setMode('javascript');

    this.aceEditor.getSession().setUseWorker(true);

    this.aceEditor.setAutoScrollEditorIntoView( true );

    this.aceEditor.getSession().setUseWrapMode(true);
    this.aceEditor.getSession().setWrapLimitRange(null, null);

    ( this.aceEditor as any ).$blockScrolling = Infinity;

    ace.config.loadModule('ace/ext/tern',  () => {

      this.aceEditor.setOptions({
        enableTern: {
          defs: ['browser', 'ecma5'],		// http://ternjs.net/doc/manual.html#option_defs
          plugins: {						// http://ternjs.net/doc/manual.html#plugins
            doc_comment: {
              fullDocs: true
            }
          },
          useWorker: true,
          switchToDoc:  (name, start)  => {
            // tern can use file switching when jump to defnition of function in another file is called
          },
          startedCb: () => {
            //once tern is enabled, it can be accessed via editor.ternServer
          },
        },
        enableSnippets: true,
        enableBasicAutocompletion: true,
        enableLiveAutocompletion: true
      });
    });

  }`

Hope this help, When do you think you'll try making it work, just to know.

Thanks in advance

**Comment by JSmithOner** [comment]
@jparismorgan I've managed to make it work just replace `this.aceEditor.session.setMode('javascript');` by `this.aceEditor.session.setMode('ace/mode/javascript');`

**Comment by jparismorgan** [comment]
Oh nice! So it is fully working? That is awesome. I will try to give it a go in my project.

**Comment by JSmithOner** [comment]
@jparismorgan yes it works and you can find ecma6 and angularjs type in tern-worker.Best.

**Comment by jparismorgan** [comment]
Hi @JSmithOner, thanks again! I'm taking a look at this now. Two issues so far:
* I had to do `loadModule('ace/tern/tern'` instead of `loadModule('ace/ext/tern',` or I would get an error that makes it look like the module cannot be found:
```
Uncaught SyntaxError: Unexpected token '<' (at ext-tern.js:1:1)
```

* After fixing that I am now having trouble importing `import 'ace-builds/src-noconflict/worker-tern'`. On page load I get:
```
Uncaught ReferenceError: f is not defined
    at Object.2 (worker-tern.js?7b14:4195:1)
    at s (worker-tern.js?7b14:4097:1)
    at eval (worker-tern.js?7b14:4099:1)
    at Object.1... (worker-tern.js?7b14:4152:1)
    at s (worker-tern.js?7b14:4097:1)
    at e (worker-tern.js?7b14:4105:38)
    at eval (worker-tern.js?7b14:4107:6)
    at eval (worker-tern.js?7b14:4062:1)
    at Object.eval (worker-tern.js?7b14:4082:2)
    at eval (worker-tern.js?7b14:21861:1)
```
which is about this line, where `acorn = undefined` so we call `f("./acorn")` and have errors:
```
module.exports = typeof acorn != "undefined" ? acorn : f("./acorn");
```


So, my questions:
* Did you see either of these errors?
* Did you have to do anything to import `acorn`?

Thanks for any help!
